### PR TITLE
Translate shims using MIR

### PR DESCRIPTION
**Pull request by arielb1**
This removes one large remaining part of old trans.

### CODE REVIEW

**Comment by KalitaAlexey** [comment]
Hi @arielb1,
Will you describe your changes?
I am interested in what you are doing.

**Comment by arielb1** [comment]
@KalitaAlexey 

Some functions in Rust, such as enum variant constructions (e.g. `Option::<u32>::Some`) and some trait methods (e.g. `<fn(u32) as Fn(u32)>::call`) are implemented through "shims". I am translating these shims more like normal functions.

**Comment by eddyb** [comment]
cc @rust-lang/compiler

**Comment by mrhota** [comment]
@arielb1 to what end? _what_ you're doing and _why_ you're doing it are equally important. Your commit messages provide information on your rationale 🙁

**Comment by jonas-schievink** [comment]
@mrhota less special-casing and less convoluted code in the compiler

**Comment by solson** [comment]
It also pulls trans-specific code out into MIR in a way that will work for alternative backends (e.g. miri, mir2wasm), reducing duplication between them.

File: src/librustc/ty/instance.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub struct Instance<'tcx> {
+    pub def: InstanceDef<'tcx>,
+    pub substs: &'tcx Substs<'tcx>,
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub enum InstanceDef<'tcx> {
+    Item(DefId),
+    // <fn() as FnTrait>::call_*
```
Comment (nikomatsakis) [inline src/librustc/ty/instance.rs]: A more explicit comment would be nice -- It's not clear to me what the `Ty<'tcx>` represents here.

File: src/librustc_mir/util/elaborate_drops.rs [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+    Open,
+}
+
+#[derive(Debug)]
+pub enum DropFlagMode {
+    Shallow,
+    Deep
+}
+
+pub trait DropElaborator<'a, 'tcx: 'a> : fmt::Debug {
```
Comment (eddyb) [inline src/librustc_mir/util/elaborate_drops.rs]: Why is this file not in `librustc_mir/transform`?

File: src/librustc/ty/context.rs [hunk 3]
```diff
@@ -423,6 +423,10 @@ pub struct GlobalCtxt<'tcx> {
     pub hir: hir_map::Map<'tcx>,
     pub maps: maps::Maps<'tcx>,
 
+    pub shim_mir_cache: RefCell<FxHashMap<ty::InstanceDef<'tcx>,
+                                          &'tcx RefCell<Mir<'tcx>>>>,
+    pub mir_shimmer: RefCell<Option<Box<ty::MirShimmer<'tcx>>>>,
+
```
Comment (eddyb) [inline src/librustc/ty/context.rs]: Can you put a `FIXME` on these to eventually integrate them better with the rest of on-demand?

File: src/librustc/ty/mod.rs [hunk 4]
```diff
@@ -2302,6 +2312,22 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         queries::mir::get(self, DUMMY_SP, did).borrow()
     }
 
+    /// Return the possibly-auto-generated MIR of a (DefId, Subst) pair.
+    pub fn instance_mir(self, instance: ty::InstanceDef<'gcx>)
```
Comment (eddyb) [inline src/librustc/ty/mod.rs]: Same `FIXME` here, I suppose.

**Review by eddyb — APPROVED** [review]
LGTM so far! r=me with those extra FIXMEs if you want to land it like this

File: src/librustc_mir/util/elaborate_drops.rs (hunk 2 above)
Comment (arielb1) [inline src/librustc_mir/util/elaborate_drops.rs]: Because it's not a MIR pass.

File: src/librustc/ty/context.rs [hunk 5]
```diff
@@ -423,6 +423,8 @@ pub struct GlobalCtxt<'tcx> {
     pub hir: hir_map::Map<'tcx>,
     pub maps: maps::Maps<'tcx>,
 
+    // This is not a ty::maps cache because it contains Tys, instead
+    // it is structured similarly to selection_cache.
```
Comment (eddyb) [inline src/librustc/ty/context.rs]: But if you have a `dep_node` method, you can use it to move this to on-demand.

**Comment by arielb1** [comment]
Now ready. r? @eddyb

**Comment by arielb1** [comment]
```
[  8%] Building CXX object lib/MC/CMakeFiles/LLVMMC.dir/MCCodeEmitter.cpp.o

DEBUG:sccache::commands: Server sent CompileStarted

make[3]: *** [lib/MC/CMakeFiles/LLVMMC.dir/MCCodeEmitter.cpp.o] Error 254

make[2]: *** [lib/MC/CMakeFiles/LLVMMC.dir/all] Error 2
```

sccache failure?

@bors retry

**Comment by arielb1** [comment]
Assertion on MSVC - confirmed locally:
```
---------------------------
Microsoft Visual C++ Runtime Library
---------------------------
Assertion failed!

Program: ...ws-msvc\stage1\bin\rustc_llvm-7d3439c525ad24fa.dll
File: C:\Rust\rust\src\llvm\lib\Transform...\InlineF...ion.cpp
Line: 394

Expression: !MemoMap.count(UselessPad) || MemoMap[UselessPad] == nullptr

For information on how your program can cause an assertion
failure, see the Visual C++ documentation on asserts

(Press Retry to debug the application - JIT must be enabled)
---------------------------
Abort   Retry   Ignore   
---------------------------
```

**Comment by eddyb** [comment]
I can't see the end of that appveyor log.

**Comment by arielb1** [comment]
sccache failure

@bors retry

**Comment by arielb1** [comment]
- #39074 
@bors retry

**Comment by arielb1** [comment]
Mac builders offline

@bors retry

**Comment by arielb1** [comment]
- #40417

@bors retry

**Comment by arielb1** [comment]
Mac test timeout?

@bors retry

**Pull request merged**